### PR TITLE
[Feature] PyTrees in replay buffers

### PR DESCRIPTION
**Pull request by vmoens**

### CODE REVIEW

File: torchrl/data/replay_buffers/storages.py [hunk 1]
```diff
@@ -347,24 +387,50 @@ def dumps(self, path):
     def loads(self, path):
         with open(path / "storage_metadata.json", "r") as file:
             metadata = json.load(file)
-        is_tensor = metadata["is_tensor"]
-        shape = metadata["shape"]
-        dtype = metadata["dtype"]
+        is_pyree = metadata["is_pyree"]
```
Comment (matteobettini) [inline torchrl/data/replay_buffers/storages.py:390]: Typo?

File: tutorials/sphinx-tutorials/rb_tutorial.py [hunk 2]
```diff
@@ -262,6 +262,50 @@ class MyData:
...
+rb = ReplayBuffer(storage=LazyMemmapStorage(100), transform=transform)
+data = {
+    "a": torch.randn(3),
+    "b": {"c": (torch.zeros(2), [torch.ones(1)])},
+    30: -torch.ones(()),  # non-string keys also work
+}
+rb.add(data)
+
+# The sample has a similar structure to the data (with a leading dimension of 10 for each tensor)
+sample = rb.sample(10)
```
Comment (matteobettini) [inline tutorials/sphinx-tutorials/rb_tutorial.py:298]: So you are sampling 10 times the same data?

Comment (vmoens) [reply tutorials/sphinx-tutorials/rb_tutorial.py:298]: yeah it's just to show the API, do you think it's confusing?

Comment (matteobettini) [reply tutorials/sphinx-tutorials/rb_tutorial.py:298]: No I think it works, maybe write explicitly this

File: docs/source/reference/data.rst [hunk 3]
```diff
@@ -22,7 +22,101 @@ widely used replay buffers:
...
+    >>> # extend with a PyTree where all tensors have the same leading dim
+    >>> rb_pytree.extend({"a": {"b": torch.randn(3), "c": [torch.zeros(3, 2), (torch.ones(3, 10),)]}})
+
+.. note:: :meth:`~torchrl.data.replay_buffers.ReplayBuffer.extend` can have an
+  ambiguous signature when dealing with lists of values, which should be interpreted
+  either as PyTree (in which case all elements in the list will be put in a slice
+  in the stored PyTree in the storage) or a list of values to add one at a time.
+  To solve this, TorchRL makes the clear-cut distinction between list and tuple:
+  a tuple will be viewed as a PyTree, a list (at the root level) will be interpreted
+  as a stack of values to add one at a time to the buffer.
```
Comment (matteobettini) [inline docs/source/reference/data.rst:85]: Good that this is documented, but I personally am not a big fan of this behavior. It seems very much a custom tailored one

File: docs/source/reference/data.rst [hunk 4]
```diff
@@ -22,7 +22,101 @@ widely used replay buffers:
...
+
+Sampling and indexing
+~~~~~~~~~~~~~~~~~~~~~
+
+Replay buffers can be indexed and sampled.
+Indexing and sampling collect data at given indices in the storage and then process them
+through a series of transforms and ``collate_fn`` that can be passed to the `__init__`
+function of the replay buffer. ``collate_fn`` comes with default values that should
+match user expectations in the majority of cases, such that you should not have
```
Comment (matteobettini) [inline docs/source/reference/data.rst:94]: Would it be possible to list the defaults?

**Review by matteobettini — COMMENTED** [review]
Some doc suggestions

File: docs/source/reference/data.rst [hunk 5]
```diff
@@ -22,7 +22,101 @@ widely used replay buffers:
...
+
+    >>> import torch
+    >>> from tensordict import TensorDict
+    >>> from torchrl.data.replay_buffers import ReplayBuffer, LazyMemmapStorage
+    >>> rb_td = ReplayBuffer(storage=LazyMemmapStorage(10), batch_size=1)  # max 10 elements stored
+    >>> rb_td.add(TensorDict({"img": torch.randint(255, (3, 64, 64), dtype=torch.unit8),
+    ...     "labels": torch.randint(100, ())}, batch_size=[]))
+    >>> rb_pytree = ReplayBuffer(storage=LazyMemmapStorage(10))  # max 10 elements stored
+    >>> # extend with a PyTree where all tensors have the same leading dim
+    >>> rb_pytree.extend({"a": {"b": torch.randn(3), "c": [torch.zeros(3, 2), (torch.ones(3, 10),)]}})
```
[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/data.rst:77]: In this case extend takes the leading dim of all tensors (as it says in the doc). I can put an assert len(rb)==3 if that helps.

I don't understand the confusion with the tutorial since it's add+sampling in that case (not extend)

[the remark this answers is not in the document]
Comment (matteobettini) [reply docs/source/reference/data.rst:77]: What if the tensors did not have a shared leading dimension or had multiple? How is that handled?

[the remark this answers is not in the document]
Comment (matteobettini) [reply docs/source/reference/data.rst:77]: Also for multiple? 

Ok I guess that was not very clear to me when reading those parapgraphs.
While for tensordict `batch_size` makes it more explicit

[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/data.rst:77]: Multiple of what?

But in general this is one of the reasons tensordict isn't just pytree on steroids, it carries metadata in a way that makes it much clearer, controlled and useful IMO.

Aside from that, if people don't want to use tensordict and they want to use pytrees instead I think it's cool we support it (since pytree is PyTorch blessed). Plus it paradoxically makes a clear case for tensordict haha 

I will clarify the docs here, I think the leading dimension is explained elsewhere but it should not require the reader to hop around the doc

[the remark this answers is not in the document]
Comment (matteobettini) [reply docs/source/reference/data.rst:77]: > Multiple of what?

> What if the tensors did not have a shared leading dimension or had _multiple_? How is that handled?

[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/data.rst:77]: yep that's fundamentally the difference between add and extend

File: docs/source/reference/data.rst (hunk 3 above)
Comment (vmoens) [reply docs/source/reference/data.rst:85]: not really, I think people think differently of list and tuples 
This is there 1) for bc-compatibility and 2) because I do think passing a list to extend is like saying "add all these elements"

File: docs/source/reference/data.rst (hunk 4 above)
Comment (vmoens) [reply docs/source/reference/data.rst:94]: it's mostly identity and it's a bit convoluted because it depends on the storage and replay buffer type. 
A list would be very confusing and hard to maintain (+ it's all private functions).
I guess the only thing users need to know is that we do our best to give back contiguous values

Comment (matteobettini) [reply docs/source/reference/data.rst:94]: maybe just mention cloning is not the default then

Comment (matteobettini) [reply docs/source/reference/data.rst:94]: jsut that accessing the data will normally not clone it (consuming more time and memory)

File: docs/source/reference/data.rst [hunk 6]
```diff
@@ -22,7 +22,101 @@ widely used replay buffers:
...
+In theory, replay buffers support any data type but we can't guarantee that each
+component will support any data type. The most crude replay buffer implementation
+is made of a :class:`~torchrl.data.replay_buffers.ReplayBuffer` base with a
+:class:`~torchrl.data.replay_buffers.ListStorage` storage. This is very inefficient
+but it will allow you to store complex data structures with non-tensor data.
+Storages in contiguous memory include :class:`~torchrl.data.replay_buffers.TensorStorage`,
+:class:`~torchrl.data.replay_buffers.LazyTensorStorage` and
+:class:`~torchrl.data.replay_buffers.LazyMemmapStorage`.
+These classes of course support :class:`~tensordict.TensorDict` data, but also
+any PyTree data structure (eg, tuples, lists, dictionaries and nested versions
```
[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/data.rst:42]: it's a private feature in torch (even if used everywhere). as of now there is no doc

File: docs/source/reference/data.rst (hunk 4 above)
Comment (vmoens) [reply docs/source/reference/data.rst:94]: right but why would people think that?
Just to understand what point we would be making
